### ISSUE TITLE
New dead_code warning for tuple struct seems wrong

### ISSUE THREAD

**Issue by taiki-e** [rust-lang/rust#119645]
I tried this code:

```rust
use std::rc::Rc;

struct Pending(Rc<()>);

fn pending() -> Pending {
    Pending(Rc::new(()))
}

fn main() {
    let _ = pending();
}
```

(The original code: https://github.com/rust-lang/futures-rs/blob/72e7e397cdfe7574e9b5d8845cafefc6e5dda70a/futures-executor/tests/local_pool.rs#L13)

I expected to see this happen: no warning or a warning suggests `Pending(PhantomData<Rc<()>>)`.

Instead, this happened:

A warning suggests `Pending(())`.

```
error: field `0` is never read
  --> futures-executor/tests/local_pool.rs:13:16
   |
13 | struct Pending(Rc<()>);
   |        ------- ^^^^^^
   |        |
   |        field in this struct
   |
   = note: `-D dead-code` implied by `-D warnings`
   = help: to override `-D warnings` add `#[allow(dead_code)]`
help: consider changing the field to be of unit type to suppress this warning while preserving the field numbering, or remove the field
   |
13 | struct Pending(());
   |                ~~
```

Changing `Rc<()>` to `()` changes auto trait implementations of `Pending`, so dead_code should not suggest such a code or should suggest both code (`PhantomData<Rc<()>>` and `()`) with explanations.

Also, if it suggests PhantomData, it should also be noted that PhantomData is not able to propagate all auto-traits (e.g., Unpin). (although this is not a problem in the above case)

https://github.com/rust-lang/futures-rs/actions/runs/7428509435/job/20215931270

### Meta

`rustc --version --verbose`:
```
rustc 1.77.0-nightly (595bc6f00 2024-01-05)
binary: rustc
commit-hash: 595bc6f00369475047538fdae1ff8cea692ac385
commit-date: 2024-01-05
host: aarch64-apple-darwin
release: 1.77.0-nightly
LLVM version: 17.0.6
```

---

This has been shown in the latest nightly (nightly-2024-01-06) since https://github.com/rust-lang/rust/pull/118297 (cc @shepmaster). IMO, it should not be changed to be warned by default with such a lint with a wrong suggestion.

**Comment by compiler-errors** [comment]
For the record, we also warn on named fields, though I feel like the suggestion shouldn't be given if the field has a `Drop` implementation at least... `PhantomData` is a good alternate suggestion to make though, I guess.

**Comment by shepmaster** [comment]
FWIW, the lint does actually check for and allow `PhantomData`, I guess the warning just doesn't mention it ([although the docs for the lint do](https://github.com/rust-lang/rust/blob/595bc6f00369475047538fdae1ff8cea692ac385/compiler/rustc_lint_defs/src/builtin.rs#L701-L703).

**Comment by shepmaster** [comment]
> we also warn on named fields

Right:

```rust
use std::rc::Rc;

struct Pending {
    a: Rc<()>,
}

fn pending() -> Pending {
    Pending { a: Rc::new(()) }
}

fn main() {
    let _ = pending();
}
```

```none
warning: field `a` is never read
 --> src/main.rs:4:5
  |
3 | struct Pending {
  |        ------- field in this struct
4 |     a: Rc<()>,
  |     ^
  |
  = note: `#[warn(dead_code)]` on by default
```

The new code is consistent with the current behavior, so I think that it's a net step forward and I'd argue about it being a *regression*.

The question becomes what do we want to do next? 

It seems like the longer prose about "PhantomData is not able to propagate all auto-traits" shouldn't go into the lint message because it feels too wordy and niche, but should certainly go into the lint docs. Or maybe it should go in the diagnostic? Maybe we get some outside diagnostics opinion?

It does seem like the named and tuple field lint text should be more similar though.

**Comment by compiler-errors** [comment]
> It does seem like the named and tuple field lint text should be more similar though.

Yeah, I agree.

> Also, if it suggests PhantomData, it should also be noted that PhantomData is not able to propagate all auto-traits (e.g., Unpin).

What do you mean by this? `PhantomData<T>` should implement (or not) all the auto traits of `T`. The trait solver specifically treats `PhantomData<T>` as if it were `PhantomData<T>(T)`:

https://github.com/rust-lang/rust/blob/595bc6f00369475047538fdae1ff8cea692ac385/compiler/rustc_trait_selection/src/solve/assembly/structural_traits.rs#L76-L77

**Comment by taiki-e** [comment]
> What do you mean by this? `PhantomData<T>` should implement (or not) all the auto traits of `T`. The trait solver specifically treats `PhantomData<T>` as if it were `PhantomData<T>(T)`:

Sorry, I thought PhantomData always implemented Unpin, but it did not (there was a PR for it but not merged: https://github.com/rust-lang/rust/pull/62786). At least at this time, it seems to propagate all (stable?) auto-traits.

**Comment by traviscross** [comment]
@rustbot labels +T-lang +I-lang-nominated

Nominating so we can discuss the proposal to mention `PhantomData` in the help message, both for field and tuple structs.

**Comment by traviscross** [comment]
@rustbot labels -I-lang-nominated

In the end, there seemed general consensus to suggest `PhantomData` only if we could do it only when removing the field would change the auto traits of the struct.

Let's go ahead and unnominate the issue.  If a PR is filed to add the suggestion to the lint in this manner, please nominate that for T-lang.

**Comment by compiler-errors** [comment]
> At most, we felt that it might be worth adding a note to the default suggestion if the removal would actually affect the auto traits of the struct (if we can tell that).

I don't believe it's possible in general to detect whether or not the auto traits of a struct are going to change in the presence of generics, normalizable associated types, and the fact that the list of auto-traits is not hard-coded into the compiler 😓

**Comment by tmandry** [comment]
Generics are another valid use case for PhantomData, making them used when they would otherwise be unused. Otherwise most uses of PhantomData to affect auto traits don't depend on generics in my experience.

My proposal would be: If we had a list of auto traits, we could ask whether the type implements each of those auto traits.

* If the answer is definitely no for any auto trait, suggest PhantomData.
* If the answer is ambiguous for any auto trait, suggest PhantomData. This ensures that auto traits will definitely be preserved under (one of) the compiler suggestions.
* If the answer is definitely yes for all auto traits, elide the PhantomData suggestion.

**Comment by clechasseur** [comment]
I see that implementations of `Drop` are mentioned... this code warns on Nightly but not on Stable:

```rust
struct Foo(&'static ());

impl Foo {
    fn new() -> Self {
        Self(&())
    }
}

impl Drop for Foo {
    fn drop(&mut self) {
        println!("Foo dropped: {:?}", self.0);
    }
}

struct Bar(Foo);

impl Bar {
    fn new() -> Self {
        Self(Foo::new())
    }
}

fn main() {
    let _ = Bar::new();
}
```

[playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=2157fe8ba36ba94223f617ce5960c7aa)

In this case, removing the field causes the `Drop` not to work properly. Is there a way to avoid the warning while keeping the field so it is correctly `Drop`ed? If I used a named field, I could prefix the field name with `_` to avoid the warning, but I don't know if there's a way to do that for fields of a tuple struct.

**Comment by glandium** [comment]
> For the record, we also warn on named fields, though I feel like the suggestion shouldn't be given if the field has a `Drop` implementation at least... `PhantomData` is a good alternate suggestion to make though, I guess.

`PhantomData` doesn't work as an alternative in the case where the field type has a `Drop` impl (e.g. in the example from https://github.com/rust-lang/rust/issues/119645#issuecomment-1905107307)

Note this is now a stable-to-beta regression.

**Comment by workingjubilee** [comment]
@clechasseur 
> In this case, removing the field causes the `Drop` not to work properly. Is there a way to avoid the warning while keeping the field so it is correctly `Drop`ed? If I used a named field, I could prefix the field name with `_` to avoid the warning, but I don't know if there's a way to do that for fields of a tuple struct.

You can simply `#[allow(dead_code)]` on the struct.
